### ISSUE TITLE
Atomics and C++

### ISSUE THREAD

**Issue by titardrew** [Netflix/vmaf#570]
Hi,
One particularly annoying thing with using libvmaf from C++ code is atomic. Suppose, I'd like to implement my own reader. It has to fill `VmafPicture`'s payload with picture data. However, to do that one has to include `picture.h` header file which includes `stdatomic.h` and uses `atomic_int` in the definition of `VmafPicture` struct. In C++ this leads to compilation errors. See [this](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=60932) and [this](https://stackoverflow.com/questions/45342776/how-to-include-c11-headers-when-compiling-c-with-gcc) for example. There's also a [proposal](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2018/p0943r1.html) to make C++ compatible with C atomics, but this is not the case now, to the best of my knowledge.
Currently, I found 2 workarounds:
1. Manually force libvmaf to compile with its `stdatomic.h` from `compat` and add 
```c
typedef int atomic_int
typedef unsigned int atomic_uint
```
before include.

2. Add:
```c
typedef int atomic_int
typedef unsigned int atomic_uint
```
or
```C
typedef std::atomic<int> atomic_int
typedef std::atomic<unsigned int> atomic_uint
```
and include `stdatomic.h` from `compat` to CMake to avoid including standard `stdatomic.h` from `picture.h`. This works, because compat stdatomic.h does nothing if included from c++ (obviously with extern "C" as well): https://github.com/Netflix/vmaf/blob/master/libvmaf/include/compat/gcc/stdatomic.h
However, I don't claim that it works properly. It just compiles well. By the way, what's the purpose of `#if !defined(__cplusplus)`? It does not seem to solve the issue, because we also have to define atomc_int in C++ code somehow.

---

The problem is that I can't come up with a proper way to include `VmafPicture` struct without a headache with atomics.
The same issue occurs if I include feature extractors to be able to get extracted features in addition to vmaf_score. Although, this seems to be illegal currently because feature extractors are not visible from header files from `include`.

**Comment by kylophone** [comment]
> ```c
> typedef int atomic_int
> typedef unsigned int atomic_uint
> ```

The above won't work, since the type needs to actually be atomic for thread safety.

Another technique might be to hide the atomic inside an anonymous struct `VmafRef`. If we did that `<stdatomic.h>` should not be required from any libvmaf API headers. What do you think about that?

**Comment by titardrew** [comment]
@kylophone,
> The above won't work, since the type needs to actually be atomic for thread safety.

Yeah, right ..
> What do you think about that?

Cool, it'd definitely work for me.
> Note that `libvmaf_rc` is not a published API yet, so it's not too late to make a breaking change like this.

Talking about the API, I'd like to have access to features produced by extractors and to be able to pool them as well. Currently, in my code, I just implemented an extension to vmaf_score_pooled so it would pool features as well. However, it is quite gross, especially because I have to include VmafContext's definition, but it is forward declared, which brings a lot of extra problems with includes and _atomics_ again. It is very convenient to have several metrics calculated and pooled at once, like e.g. SSIM and PSNR along with (or without) vmaf_score. I think it is not that hard to implement it in the API (though there're some design questions). What do you think about the feature?

**Comment by kylophone** [comment]
Fixed by #623.
